projects/uriparser/uri_dissect_query_malloc_fuzzer.cc

```diff
@@ -0,0 +1,76 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Fuzz UriQuery.c:
+//   uriDissectQueryMallocA
+//   uriComposeQueryA
+
+#include <cstddef>
+#include <cstdint>
+#include <string>
+#include <utility>
+#include <vector>
+#include <fuzzer/FuzzedDataProvider.h>
+
+using std::string;
+using std::make_pair;
+#include "uriparser/include/uriparser/Uri.h"
+#include "uriparser/include/uriparser/UriBase.h"
+
+std::vector<std::pair<std::string, std::string>> ToVector(
+    UriQueryListA *query_list) {
+  std::vector<std::pair<std::string, std::string>> result;
+  if (query_list == nullptr) return result;
+  for (UriQueryListA *entry = query_list; entry != nullptr;
+       entry = entry->next) {
+    // The value can be a nullptr.
+    result.push_back(std::make_pair(
+        entry->key, entry->value == nullptr ? "null" : entry->value));
+  }
+  uriFreeQueryListA(query_list);
+  return result;
+}
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
+  FuzzedDataProvider stream(data, size);
+  uint16_t maxSize = stream.ConsumeIntegral<uint16_t>();
```

Comment:
I double-checked the URIParser [documentation](https://uriparser.github.io/doc/api/latest/Uri_8h.html), and it looks like there are two ways to avoid having to create `maxSize` yourself.  Either use [`uriComposeQueryCharsRequiredA`](https://uriparser.github.io/doc/api/latest/Uri_8h.html#a45b2899b4cdeeb8d229543ea3edfd872) to calculate the appropriate `maxSize`, or use [`uriComposeQueryMallocA`](https://uriparser.github.io/doc/api/latest/Uri_8h.html#a6bb58a31d3902e681b1747a55da27015) to have the library allocate memory for you.

Reply:
`uriComposeQueryCharsRequiredA` works, looks like we don't need to use `FuzzedDataProvider` anymore.